File: forgerock-auth/src/test/java/org/forgerock/android/auth/callback/NumberAttributeInputCallbackTest.java
/*
 * Copyright (c) 2019 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth.callback;

import org.assertj.core.api.Assertions;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class NumberAttributeInputCallbackTest {

    @Test
    public void basicTest() throws JSONException {
        JSONObject raw = new JSONObject("{\n" +
                "            \"type\": \"NumberAttributeInputCallback\",\n" +
                "            \"output\": [\n" +
                "                {\n" +
                "                    \"name\": \"name\",\n" +
                "                    \"value\": \"age\"\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"prompt\",\n" +
                "                    \"value\": \"Age\"\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"required\",\n" +
                "                    \"value\": true\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"policies\",\n" +
                "                    \"value\": {\n" +
                "                        \"policyRequirements\": [\n" +
                "                            \"VALID_TYPE\"\n" +
                "                        ],\n" +
                "                        \"fallbackPolicies\": null,\n" +
                "                        \"name\": \"age\",\n" +
                "                        \"policies\": [\n" +
                "                            {\n" +
                "                                \"policyRequirements\": [\n" +
                "                                    \"VALID_TYPE\"\n" +
                "                                ],\n" +
                "                                \"policyId\": \"valid-type\",\n" +
                "                                \"params\": {\n" +
                "                                    \"types\": [\n" +
                "                                        \"number\"\n" +
                "                                    ]\n" +
                "                                }\n" +
                "                            }\n" +
                "                        ],\n" +
                "                        \"conditionalPolicies\": null\n" +
                "                    }\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"failedPolicies\",\n" +
                "                    \"value\": []\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"validateOnly\",\n" +
                "                    \"value\": false\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"value\",\n" +
                "                    \"value\": null\n" +
                "                }\n" +
                "            ],\n" +
                "            \"input\": [\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1\",\n" +
                "                    \"value\": null\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1validateOnly\",\n" +
                "                    \"value\": false\n" +
                "                }\n" +
                "            ]\n" +
                "        }");
        NumberAttributeInputCallback callback = new NumberAttributeInputCallback(raw, 0);
        Assertions.assertThat(callback.getName()).isEqualTo("age");
        Assertions.assertThat(callback.getPrompt()).isEqualTo("Age");
        Assertions.assertThat(callback.isRequired()).isTrue();
        Assertions.assertThat(callback.getPolicies().getString("name")).isEqualTo("age");
        Assertions.assertThat(callback.getFailedPolicies()).isEmpty();
        Assertions.assertThat(callback.getValidateOnly()).isFalse();
        Assertions.assertThat(callback.getValue()).isNull();


    }

    @Test
    public void basicInteger() throws JSONException {
        JSONObject raw = new JSONObject("{\n" +
                "            \"type\": \"NumberAttributeInputCallback\",\n" +
                "            \"output\": [\n" +
                "                {\n" +
                "                    \"name\": \"name\",\n" +
                "                    \"value\": \"age\"\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"prompt\",\n" +
                "                    \"value\": \"Age\"\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"required\",\n" +
                "                    \"value\": true\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"policies\",\n" +
                "                    \"value\": {\n" +
                "                        \"policyRequirements\": [\n" +
                "                            \"VALID_TYPE\"\n" +
                "                        ],\n" +
                "                        \"fallbackPolicies\": null,\n" +
                "                        \"name\": \"age\",\n" +
                "                        \"policies\": [\n" +
                "                            {\n" +
                "                                \"policyRequirements\": [\n" +
                "                                    \"VALID_TYPE\"\n" +
                "                                ],\n" +
                "                                \"policyId\": \"valid-type\",\n" +
                "                                \"params\": {\n" +
                "                                    \"types\": [\n" +
                "                                        \"number\"\n" +
                "                                    ]\n" +
                "                                }\n" +
                "                            }\n" +
                "                        ],\n" +
                "                        \"conditionalPolicies\": null\n" +
                "                    }\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"failedPolicies\",\n" +
                "                    \"value\": []\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"validateOnly\",\n" +
                "                    \"value\": false\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"value\",\n" +
                "                    \"value\": 123\n" +
                "                }\n" +
                "            ],\n" +
                "            \"input\": [\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1\",\n" +
                "                    \"value\": null\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1validateOnly\",\n" +
                "                    \"value\": false\n" +
                "                }\n" +
                "            ]\n" +
                "        }");
        NumberAttributeInputCallback callback = new NumberAttributeInputCallback(raw, 0);
        Assertions.assertThat(callback.getValue()).isEqualTo(123);
    }

    @Test
    public void basicDecimal() throws JSONException {
        JSONObject raw = new JSONObject("{\n" +
                "            \"type\": \"NumberAttributeInputCallback\",\n" +
                "            \"output\": [\n" +
                "                {\n" +
                "                    \"name\": \"name\",\n" +
                "                    \"value\": \"age\"\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"prompt\",\n" +
                "                    \"value\": \"Age\"\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"required\",\n" +
                "                    \"value\": true\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"policies\",\n" +
                "                    \"value\": {\n" +
                "                        \"policyRequirements\": [\n" +
                "                            \"VALID_TYPE\"\n" +
                "                        ],\n" +
                "                        \"fallbackPolicies\": null,\n" +
                "                        \"name\": \"age\",\n" +
                "                        \"policies\": [\n" +
                "                            {\n" +
                "                                \"policyRequirements\": [\n" +
                "                                    \"VALID_TYPE\"\n" +
                "                                ],\n" +
                "                                \"policyId\": \"valid-type\",\n" +
                "                                \"params\": {\n" +
                "                                    \"types\": [\n" +
                "                                        \"number\"\n" +
                "                                    ]\n" +
                "                                }\n" +
                "                            }\n" +
                "                        ],\n" +
                "                        \"conditionalPolicies\": null\n" +
                "                    }\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"failedPolicies\",\n" +
                "                    \"value\": []\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"validateOnly\",\n" +
                "                    \"value\": false\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"value\",\n" +
                "                    \"value\": 23.2\n" +
                "                }\n" +
                "            ],\n" +
                "            \"input\": [\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1\",\n" +
                "                    \"value\": null\n" +
                "                },\n" +
                "                {\n" +
                "                    \"name\": \"IDToken1validateOnly\",\n" +
                "                    \"value\": false\n" +
                "                }\n" +
                "            ]\n" +
                "        }");
        NumberAttributeInputCallback callback = new NumberAttributeInputCallback(raw, 0);
        Assertions.assertThat(callback.getValue()).isEqualTo(23.2);
    }
}

File: forgerock-auth/src/test/java/org/forgerock/android/auth/callback/NumberAttributeInputCallbackTest.kt
/*
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 *  This software may be modified and distributed under the terms
 *  of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth.callback

import androidx.test.ext.junit.runners.AndroidJUnit4
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.*
import org.json.JSONException
import org.json.JSONObject
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class NumberAttributeInputCallbackTest {
    @Test
    @Throws(JSONException::class)
    fun basicTest() {
        val raw = JSONObject("""{
            "type": "NumberAttributeInputCallback",
            "output": [
                {
                    "name": "name",
                    "value": "age"
                },
                {
                    "name": "prompt",
                    "value": "Age"
                },
                {
                    "name": "required",
                    "value": true
                },
                {
                    "name": "policies",
                    "value": {
                        "policyRequirements": [
                            "VALID_TYPE"
                        ],
                        "fallbackPolicies": null,
                        "name": "age",
                        "policies": [
                            {
                                "policyRequirements": [
                                    "VALID_TYPE"
                                ],
                                "policyId": "valid-type",
                                "params": {
                                    "types": [
                                        "number"
                                    ]
                                }
                            }
                        ],
                        "conditionalPolicies": null
                    }
                },
                {
                    "name": "failedPolicies",
                    "value": []
                },
                {
                    "name": "validateOnly",
                    "value": false
                },
                {
                    "name": "value",
                    "value": null
                }
            ],
            "input": [
                {
                    "name": "IDToken1",
                    "value": null
                },
                {
                    "name": "IDToken1validateOnly",
                    "value": false
                }
            ]
        }""")
        val callback = NumberAttributeInputCallback(raw, 0)
        assertThat(callback.name).isEqualTo("age")
        assertThat(callback.prompt).isEqualTo("Age")
        assertThat(callback.isRequired).isTrue
        assertThat(callback.policies.getString("name")).isEqualTo("age")
        assertThat(callback.failedPolicies).isEmpty()
        assertThat(callback.validateOnly).isFalse
        @Suppress("USELESS_CAST")
        assertThat(callback.value as Double?).isNull()
    }

    @Test
    @Throws(JSONException::class)
    fun basicInteger() {
        val raw = JSONObject("""{
            "type": "NumberAttributeInputCallback",
            "output": [
                {
                    "name": "name",
                    "value": "age"
                },
                {
                    "name": "prompt",
                    "value": "Age"
                },
                {
                    "name": "required",
                    "value": true
                },
                {
                    "name": "policies",
                    "value": {
                        "policyRequirements": [
                            "VALID_TYPE"
                        ],
                        "fallbackPolicies": null,
                        "name": "age",
                        "policies": [
                            {
                                "policyRequirements": [
                                    "VALID_TYPE"
                                ],
                                "policyId": "valid-type",
                                "params": {
                                    "types": [
                                        "number"
                                    ]
                                }
                            }
                        ],
                        "conditionalPolicies": null
                    }
                },
                {
                    "name": "failedPolicies",
                    "value": []
                },
                {
                    "name": "validateOnly",
                    "value": false
                },
                {
                    "name": "value",
                    "value": 123
                }
            ],
            "input": [
                {
                    "name": "IDToken1",
                    "value": null
                },
                {
                    "name": "IDToken1validateOnly",
                    "value": false
                }
            ]
        }""")
        val callback = NumberAttributeInputCallback(raw, 0)
        assertThat(callback.value).isEqualTo(123.0)
    }

    @Test
    @Throws(JSONException::class)
    fun basicDecimal() {
        val raw = JSONObject("""{
            "type": "NumberAttributeInputCallback",
            "output": [
                {
                    "name": "name",
                    "value": "age"
                },
                {
                    "name": "prompt",
                    "value": "Age"
                },
                {
                    "name": "required",
                    "value": true
                },
                {
                    "name": "policies",
                    "value": {
                        "policyRequirements": [
                            "VALID_TYPE"
                        ],
                        "fallbackPolicies": null,
                        "name": "age",
                        "policies": [
                            {
                                "policyRequirements": [
                                    "VALID_TYPE"
                                ],
                                "policyId": "valid-type",
                                "params": {
                                    "types": [
                                        "number"
                                    ]
                                }
                            }
                        ],
                        "conditionalPolicies": null
                    }
                },
                {
                    "name": "failedPolicies",
                    "value": []
                },
                {
                    "name": "validateOnly",
                    "value": false
                },
                {
                    "name": "value",
                    "value": 23.2
                }
            ],
            "input": [
                {
                    "name": "IDToken1",
                    "value": null
                },
                {
                    "name": "IDToken1validateOnly",
                    "value": false
                }
            ]
        }""")
        val callback = NumberAttributeInputCallback(raw, 0)
        assertThat(callback.value).isEqualTo(23.2)
    }
}